drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/PaymentMethodsListDelegate.kt

```diff
@@ -0,0 +1,198 @@
+    private val amount: Amount
+) : ComponentAvailableCallback<Configuration> {
+
+    private val _paymentMethodsFlow = MutableStateFlow<List<PaymentMethodListItem>>(emptyList())
+    internal val paymentMethodsFlow: StateFlow<List<PaymentMethodListItem>> = _paymentMethodsFlow
+
+    private lateinit var storedPaymentMethodsList: MutableList<StoredPaymentMethodModel>
+    private var checkPaymentMethodsMap: HashMap<PaymentMethod, Boolean> = hashMapOf()
+
+    init {
+        setupPaymentMethods(paymentMethods)
+    }
+
+    internal fun getPaymentMethod(index: Int): PaymentMethod {
+        return paymentMethods[index]
+    }
+
+    private fun setupPaymentMethods(paymentMethods: List<PaymentMethod>) {
+        paymentMethods.forEach { paymentMethod ->
+            val type = paymentMethod.type ?: throw IllegalStateException("PaymentMethod type is null")
+
+            when {
+                PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type) -> {
+                    Logger.d(TAG, "Supported payment method: $type")
+                    checkPaymentMethodAvailability(application, paymentMethod, dropInConfiguration, amount, this)
+                }
+                PaymentMethodTypes.UNSUPPORTED_PAYMENT_METHODS.contains(type) -> {
+                    Logger.e(TAG, "PaymentMethod not yet supported - $type")
+                    checkPaymentMethodsMap[paymentMethod] = false
+                }
+                else -> {
+                    Logger.d(TAG, "No availability check required - $type")
+                    checkPaymentMethodsMap[paymentMethod] = true
+                }
+            }
+        }
+        checkIfListReady()
+    }
+
+    override fun onAvailabilityResult(isAvailable: Boolean, paymentMethod: PaymentMethod, config: Configuration?) {
+        Logger.d(TAG, "onAvailabilityResult - ${paymentMethod.type}: $isAvailable")
+        checkPaymentMethodsMap[paymentMethod] = isAvailable
+        checkIfListReady()
+    }
+
+    private fun checkIfListReady() {
+        if (paymentMethods.size == checkPaymentMethodsMap.size) {
+            populatePaymentMethods()
+        }
+    }
+
+    private fun populatePaymentMethods() {
+        val paymentMethods = mutableListOf<PaymentMethodListItem>().apply {
+            // gift cards
+            val giftCardsList: List<GiftCardPaymentMethodModel> =
+                order?.paymentMethods?.mapToGiftCardPaymentMethodModel().orEmpty()
+            if (giftCardsList.isNotEmpty()) {
+                add(PaymentMethodHeader(PaymentMethodHeader.TYPE_GIFT_CARD_HEADER))
+                addAll(giftCardsList)
+            }
+            // payment notes
+            order?.remainingAmount?.let { remainingAmount ->
+                val value = CurrencyUtils.formatAmount(remainingAmount, dropInConfiguration.shopperLocale)
+                add(
+                    PaymentMethodNote(application.getString(R.string.checkout_giftcard_pay_remaining_amount, value))
```

Comment:
It's a bad practice to use context outside of view related classes. It would be better to pass to string id int and let the fragment resolve it

Reply:
the problem is we still need to call context for this:
```
checkPaymentMethodAvailability(application, paymentMethod, dropInConfiguration, amount, this)
```

Reply:
Ah didn't see that, that's unfortunate